Plutarch/DataRepr/Internal.hs

```diff
@@ -180,7 +184,55 @@ type family PUnLabel (a :: PLabeledType) :: PType where
  This is how most data structures are stored on-chain.
 -}
 type PDataSum :: [[PLabeledType]] -> PType
-data PDataSum (defs :: [[PLabeledType]]) (s :: S)
+data PDataSum (defs :: [[PLabeledType]]) (s :: S) where 
+  DSZ :: (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) -> Term s out) 
+      -> PDataSum (x ': xs) s
+
+  DSS :: PDataSum xs s -> PDataSum (x ': xs) s
+
+-- the length functions in sop-core gives the actual length, we want length-1
+adjustedLength :: forall x xs proxy. SListI (x:xs) => proxy (x:xs) -> Integer 
+adjustedLength proxy = fromIntegral (lengthSList proxy) - 1  
+
+instance SListI (def:defs) => PlutusType (PDataSum (def:defs)) where
+  type PInner (PDataSum (def:defs)) _ = PData 
+
+  pcon' :: forall 
+           {s :: S} 
+           {b :: PType}
+         . PDataSum (def:defs) s 
+        -> Term s (PInner (PDataSum (def:defs)) b)
+  pcon'  = \case 
+        DSZ f ->  punsafeBuiltin PLC.ConstrData # (pconstant @PInteger $ adjustedLength $ Proxy @(def:defs)) # (f id) -- rewrite this with pconstrBuiltin?
+        DSS rest -> case rest of 
+          DSZ f -> punsafeBuiltin PLC.ConstrData # (pconstant @PInteger $ adjustedLength $ Proxy @(def:defs)) # (f id)
+          dss@(DSS _) -> pcon' dss  
```

Comment:
Isn't the index in this case incorrect?

Reply:
Ugh yeah, sorry about that. I made all the changes in a rush and missed that. I think it should be 

``` haskell 
        DSS rest -> case rest of 
          DSZ f -> punsafeBuiltin PLC.ConstrData # (pconstant @PInteger $ adjustedLength $ Proxy @(defs)) # (f id)
          dss@(DSS _) -> pcon' dss  
```